pyomo/solver/ipopt.py

```diff
@@ -0,0 +1,517 @@
+                basename + '.row', 'w'
+            ) as row_file, open(basename + '.col', 'w') as col_file:
+                nl_info = self._writer.write(
+                    model,
+                    nl_file,
+                    row_file,
+                    col_file,
+                    symbolic_solver_labels=config.symbolic_solver_labels,
+                )
+            # Get a copy of the environment to pass to the subprocess
+            env = os.environ.copy()
+            if nl_info.external_function_libraries:
+                if env.get('AMPLFUNC'):
+                    nl_info.external_function_libraries.append(env.get('AMPLFUNC'))
+                env['AMPLFUNC'] = "\n".join(nl_info.external_function_libraries)
+            symbol_map = self._symbol_map = SymbolMap()
+            labeler = NumericLabeler('component')
+            for v in nl_info.variables:
+                symbol_map.getSymbol(v, labeler)
+            for c in nl_info.constraints:
+                symbol_map.getSymbol(c, labeler)
+            # Write the opt_file, if there should be one; return a bool to say
+            # whether or not we have one (so we can correctly build the command line)
+            opt_file = self._write_options_file(
+                filename=basename, options=self.solver_options
+            )
+            # Call ipopt - passing the files via the subprocess
+            cmd = self._create_command_line(
+                basename=basename, config=config, opt_file=opt_file
+            )
+            # this seems silly, but we have to give the subprocess slightly longer to finish than
+            # ipopt
+            if config.time_limit is not None:
+                timeout = config.time_limit + min(
+                    max(1.0, 0.01 * config.time_limit), 100
+                )
+            else:
+                timeout = None
+
+            ostreams = [io.StringIO()]
+            if config.tee:
+                ostreams.append(sys.stdout)
+            else:
+                ostreams.append(
+                    LogStream(
+                        level=config.log_level, logger=config.solver_output_logger
+                    )
+                )
+            with TeeStream(*ostreams) as t:
+                process = subprocess.run(
+                    cmd,
+                    timeout=timeout,
+                    env=env,
+                    universal_newlines=True,
+                    stdout=t.STDOUT,
+                    stderr=t.STDERR,
+                )
+                # This is the stuff we need to parse to get the iterations
+                # and time
+                iters, ipopt_time_nofunc, ipopt_time_func = self._parse_ipopt_output(
+                    ostreams[0]
+                )
+
+            if process.returncode != 0:
+                results.termination_condition = TerminationCondition.error
+                results.solution_loader = SolutionLoader(None, None, None, None)
```

Comment:
Don't need `None`s - they all default to that anyway.

Reply:
Right now, these don't have default values, so the `None`s are needed.

Reply:
You're right. In an earlier iteration, I had the default set in `SolverConfig` and then undid it because I didn't want the type to be `SolutionLoader` - I wanted it to be `SolutionLoaderBase`.